### PR TITLE
AI Test Toolkit: Update telemetry

### PR DESCRIPTION
**Pull request by t-prda**
### Feature Telemetry:

1. **Discovered**
   - Whenever the page is opened

2. **Set Up**
   - Whenever running the start suite or start suite line
   - Custom dimensions:
     - Run ID (only for suite header)
     - Version
     - Number of test lines (only for suite header)

3. **Used**
   - Test suite is finished running
   - Custom dimensions:
     - Run ID (only for suite header)
     - Version
     - Number of tests executed
     - Number of tests passed
     - Total time taken

4. **Log Usage**
   - Dataset imported
     - CD: Number of lines imported, Filetype

#### Work Item(s)
Fixes [AB#533358](https://dynamicssmb2.visualstudio.com/1fcb79e7-ab07-432a-a3c6-6cf5a88ba4a5/_workitems/edit/533358)

### CODE REVIEW

**Review by BardurKnudsen — DISMISSED** [review]

**Review by stkillen — APPROVED** [review]

File: src/Tools/AI Test Toolkit/src/AITCommandLineCard.Page.al [hunk 1]
```diff
@@ -208,8 +209,12 @@ page 149042 "AIT CommandLine Card"
     trigger OnOpenPage()
     var
         EnvironmentInformation: Codeunit "Environment Information";
+        FeatureTelemetry: Codeunit "Feature Telemetry";
+        AITTestSuiteMgt: Codeunit "AIT Test Suite Mgt.";
     begin
         EnableActions := (EnvironmentInformation.IsSaaS() and EnvironmentInformation.IsSandbox()) or EnvironmentInformation.IsOnPrem();
+        if EnableActions then
+            FeatureTelemetry.LogUptake('', AITTestSuiteMgt.GetFeatureName(), Enum::"Feature Uptake Status"::Discovered);
```
Comment (darjoo) [inline src/Tools/AI Test Toolkit/src/AITCommandLineCard.Page.al]: Missing tag

File: src/Tools/AI Test Toolkit/src/TestSuite/AITTestSuite.Page.al [hunk 2]
```diff
@@ -269,8 +270,11 @@ page 149031 "AIT Test Suite"
     trigger OnOpenPage()
     var
         EnvironmentInformation: Codeunit "Environment Information";
+        FeatureTelemetry: Codeunit "Feature Telemetry";
     begin
         EnableActions := (EnvironmentInformation.IsSaaS() and EnvironmentInformation.IsSandbox()) or EnvironmentInformation.IsOnPrem();
+        if EnableActions then
+            FeatureTelemetry.LogUptake('', AITTestSuiteMgt.GetFeatureName(), Enum::"Feature Uptake Status"::Discovered);
```
Comment (darjoo) [inline src/Tools/AI Test Toolkit/src/TestSuite/AITTestSuite.Page.al]: Missing tag

File: src/Tools/AI Test Toolkit/src/TestSuite/AITTestSuiteMgt.Codeunit.al [hunk 3]
```diff
@@ -64,23 +65,36 @@ codeunit 149034 "AIT Test Suite Mgt."
...
     begin
-        if UpdateSuiteVersion then begin
+        if not IsExecutedFromTestSuiteHeader then begin
             AITTestSuite.Get(AITTestMethodLine."Test Suite Code");
             AITTestSuite.Version += 1;
             AITTestSuite.Modify(true);
+
+            // Log the feature telemetry when executed from the test method line
+            TelemetryCustomDimensions.Add('Version', Format(AITTestSuite.Version));
+            FeatureTelemetry.LogUptake('', GetFeatureName(), Enum::"Feature Uptake Status"::"Set up", TelemetryCustomDimensions);
```
Comment (darjoo) [inline src/Tools/AI Test Toolkit/src/TestSuite/AITTestSuiteMgt.Codeunit.al]: Missing tag

File: src/Tools/AI Test Toolkit/src/TestSuite/AITTestSuiteMgt.Codeunit.al [hunk 4]
```diff
@@ -94,6 +108,14 @@ codeunit 149034 "AIT Test Suite Mgt."
             AITTestMethodLine.Validate(Status, AITTestMethodLine.Status::Completed);
             AITTestMethodLine.Modify(true);
             Commit();
+
+            // Log the feature telemetry when execution from the test method line has completed
+            if not IsExecutedFromTestSuiteHeader then begin
+                AITTestMethodLine.SetRange("Version Filter", AITTestSuite.Version);
+                AITTestMethodLine.CalcFields("No. of Tests Executed", "No. of Tests Passed", "Total Duration (ms)");
+                TelemetryCustomDimensions := GetFeatureUsedInsights(EmptyGuid, AITTestSuite.Version, AITTestMethodLine."No. of Tests Executed", AITTestMethodLine."No. of Tests Passed", AITTestMethodLine."Total Duration (ms)");
+                FeatureTelemetry.LogUptake('', GetFeatureName(), Enum::"Feature Uptake Status"::Used, TelemetryCustomDimensions);
```
Comment (darjoo) [inline src/Tools/AI Test Toolkit/src/TestSuite/AITTestSuiteMgt.Codeunit.al]: Missing tag

File: src/Tools/AI Test Toolkit/src/TestSuite/AITTestSuiteMgt.Codeunit.al [hunk 5]
```diff
@@ -189,31 +211,22 @@ codeunit 149034 "AIT Test Suite Mgt."
...
-                Session.LogMessage('0000DHT', AITRunCancelledLbl, Verbosity::Normal, DataClassification::SystemMetadata, TelemetryScope::All, TelemetryCustomDimensions);
-        end;
         AITTestSuite.Modify(true);
         Commit();
+
+        // Log feature telemetry when execution from the test suite header has completed
+        if AITTestSuite.Status = AITTestSuite.Status::Completed then begin
+            AITTestSuite.CalcFields("No. of Tests Executed", "No. of Tests Passed", "Total Duration (ms)");
+            TelemetryCustomDimensions := GetFeatureUsedInsights(AITTestSuite.RunID, AITTestSuite.Version, AITTestSuite."No. of Tests Executed", AITTestSuite."No. of Tests Passed", AITTestSuite."Total Duration (ms)");
+            FeatureTelemetry.LogUptake('', FeatureNameLbl, Enum::"Feature Uptake Status"::Used, TelemetryCustomDimensions);
```
Comment (darjoo) [inline src/Tools/AI Test Toolkit/src/TestSuite/AITTestSuiteMgt.Codeunit.al]: Missing tag

File: src/Tools/Test Framework/Test Runner/src/DataDrivenTest/DataInputs/TestInputsManagement.Codeunit.al [hunk 6]
```diff
@@ -109,18 +109,30 @@ codeunit 130458 "Test Inputs Management"
...
+        if FileName.EndsWith(JsonlFileExtensionTxt) then begin
+            FileType := JsonlFileExtensionTxt;
             ParseDataInputsJsonl(TestInputInStream, TestInputGroup);
+        end;
+
+        // Log telemetry for the number of lines imported and the file type
+        TestInput.SetRange("Test Input Group Code", TestInputGroup.Code);
+        TelemetryCD.Add('File Type', FileType);
+        TelemetryCD.Add('No. of entries', Format(TestInput.Count()));
+        Session.LogMessage('', 'Data Driven Test: Test Input Imported', Verbosity::Normal, DataClassification::SystemMetadata, TelemetryScope::All, TelemetryCD);
```
Comment (darjoo) [inline src/Tools/Test Framework/Test Runner/src/DataDrivenTest/DataInputs/TestInputsManagement.Codeunit.al]: Missing tag

**Review by darjoo — APPROVED** [review]

**Review by stkillen — APPROVED** [review]

**Pull request merged**